RecoHI/HiJetAlgos/plugins/HiPuRhoProducer.cc

```diff
@@ -0,0 +1,645 @@
+      puPtMin_(iConfig.getParameter<double>("puPtMin")),
+      radiusPU_(iConfig.getParameter<double>("radiusPU")),
+      rParam_(iConfig.getParameter<double>("rParam")),
+      src_(iConfig.getParameter<edm::InputTag>("src")),
+      towSigmaCut_(iConfig.getParameter<double>("towSigmaCut")) {
+  input_candidateview_token_ = consumes<reco::CandidateView>(src_);
+
+  //register your products
+  produces<std::vector<double>>("mapEtaEdges");
+  produces<std::vector<double>>("mapToRho");
+  produces<std::vector<double>>("mapToRhoMedian");
+  produces<std::vector<double>>("mapToRhoExtra");
+  produces<std::vector<double>>("mapToRhoM");
+  produces<std::vector<int>>("mapToNTow");
+  produces<std::vector<double>>("mapToTowExcludePt");
+  produces<std::vector<double>>("mapToTowExcludePhi");
+  produces<std::vector<double>>("mapToTowExcludeEta");
+}
+
+// ------------ method called to produce the data  ------------
+void HiPuRhoProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  setupGeometryMap(iEvent, iSetup);
+
+  inputs_.clear();
+  fjInputs_.clear();
+  fjJets_.clear();
+
+  // Get the particletowers
+  edm::Handle<reco::CandidateView> inputsHandle;
+  iEvent.getByToken(input_candidateview_token_, inputsHandle);
+
+  for (size_t i = 0; i < inputsHandle->size(); ++i)
+    inputs_.push_back(inputsHandle->ptrAt(i));
+
+  fjInputs_.reserve(inputs_.size());
+  inputTowers();
+  fjOriginalInputs_ = fjInputs_;
+
+  calculatePedestal(fjInputs_);
+  subtractPedestal(fjInputs_);
+
+  fjJetDefinition_ = std::make_shared<fastjet::JetDefinition>(fastjet::antikt_algorithm, rParam_);
+  fjClusterSeq_ = std::make_shared<fastjet::ClusterSequence>(fjInputs_, *fjJetDefinition_);
+  fjJets_ = fastjet::sorted_by_pt(fjClusterSeq_->inclusive_jets(puPtMin_));
+
+  etaEdgeLow_.clear();
+  etaEdgeHi_.clear();
+  etaEdges_.clear();
+
+  rho_.clear();
+  rhoExtra_.clear();
+  rhoM_.clear();
+  nTow_.clear();
+
+  towExcludePt_.clear();
+  towExcludePhi_.clear();
+  towExcludeEta_.clear();
+
+  std::vector<fastjet::PseudoJet> orphanInput;
+  calculateOrphanInput(orphanInput);
+  calculatePedestal(orphanInput);
+  putRho(iEvent, iSetup);
+}
+
+void HiPuRhoProducer::inputTowers() {
+  auto inBegin = inputs_.begin();
+  auto inEnd = inputs_.end();
+  for (auto i = inBegin; i != inEnd; ++i) {
+    reco::CandidatePtr input = *i;
+
+    if (edm::isNotFinite(input->pt()))
+      continue;
+    if (input->pt() < 100 * std::numeric_limits<double>::epsilon())
+      continue;
+
+    fjInputs_.push_back(fastjet::PseudoJet(input->px(), input->py(), input->pz(), input->energy()));
+    fjInputs_.back().set_user_index(i - inBegin);
+  }
+}
+
+void HiPuRhoProducer::setupGeometryMap(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  LogDebug("PileUpSubtractor") << "The subtractor setting up geometry...\n";
+
+  if (geo_ == nullptr) {
+    edm::ESHandle<CaloGeometry> pG;
+    iSetup.get<CaloGeometryRecord>().get(pG);
```

Comment:
Please use `esConsumes()` and `ESGetToken` to access EventSetup products https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#In_ED_module